design/build-clusters.md

```diff
@@ -0,0 +1,140 @@
+
+### Yaml 
+
+```yml
+shared:
+    environment:
+    NODE_ENV: test
+    settings:
+        email:
+    addresses: [test@email.com, test2@email.com]
+    statuses: [SUCCESS, FAILURE]
+    annotations:
+        buildCluster: iOS
+jobs:
+    main:
+        requires: [~pr, ~commit]
+        sourcePaths: ["src/app/", "screwdriver.yaml"]
+        image: node:6
+        steps:
+    - init: npm install
+    - test: npm test
+    publish:
+    requires: main
+    image: node:6
+    steps:
+        - publish: npm publish
+    ...
+```   
+ 
+### New table for build cluster details
+Table: `buildClusters`
+
+Columns:
+
+| Name | Type | Nullable | Primary Key | Unique | Description
+| --- | --- | --- | --- | --- | --- |
+| `id` | integer | no | yes | yes | |
+| `name` | text (100) | no | no | yes | |
+| `scmOrganization` | text(200) | no | no | no | |
+| `scmContext` | text(200) | no | no | no | |
+| `isActive` | boolean | no | no | no | *0-false or 1-true* |
+| `authKeyFileName` | text(100) | no | no | yes | *encrypt and store the public key filename* |
+| `signOptionFileName` | text(100) | no | no | yes | *encrypt and store signOption filename { issuer: , subject: , audience: , expires: , algorithm: }* |
+| `queue` | text(100) | no | no | no | |
+| `managedBy` | text(50) | no | no | no | cluster managed by *screwdriver or external* |
+
+
+### Cache 
+
+	Use Redis HMSET to cache all active records from `buildClusters` table. Redis should be updated for any changes in `buildClusters` table.
+		HMSET <cluster_name> authKeyName "filename" signOption "{value}" queue "value" managedBy "screwdriver"
+		HGET <cluster_name> authKeyName
+		HGET <cluster_name> signOption
+
+### Queue
+
+Redis queue (Master) - It will hold queues for all build clusters. 
+
+Redis queue (Build Cluster) - It will hold queue wrto the build cluster.
+
+
+### Below listed apis need to be built to manage the cluster details
+
+| Method | url |
+| --- | --- |
+| `PUT` | ` /v4/buildClusters { } ` |
+| `PUT` | `	/v4/buildClusters {name} ` |
+| `GET` | `	/v4/buildClusters ` |
+| `GET` | `	/v4/buildClusters {name} ` |
+| `DELETE` | ` /v4/buildClusters {name} ` |
+
+
+## Flow
+### Screwdriver API to Scheduler service
+
+	1. Screwdriver UI / PR commit / Merge triggers new build via Screwdriver API
+	2. Screwdriver API inturn calls Scheduler service with appropriate build details to schedule a job
+	3. Authentication via JWT token (refer Authorization section)
+	4. Scheduler service queries `buildClusters` table for active records with cluster name from build info and validate if job can be scheduled in apropriate buildCluster queue
+	5. one (or) more record exist, then assign job to the queue identified by generating a random number within given boundaries, which is the returned list size of records
```

Comment:
just to double check. more records are for screwdriver clusters right? external should always get one or they can also provide multiple?

Reply:
external can provide multiple as well

Reply:
based on our discussion, lets keep cluster name unique and removed queue